Paper: Evaluating probabilistic forecasters with sktime - an easy-to-use, highly configurable framework for reproducible science (#915)
scipy-conference/scipy_proceedings: papers/benchmarking/main.tex

\end{abstract}

\section{Introduction}\label{introduction}
Making probabilistic forecasts is challenging, and evaluating probabilistic forecasts, or the algorithms that produce them, is even more difficult. 

A significant body of literature focuses on developing robust meta-methodologies for evaluation. This includes evaluation metrics such as the Continuous Ranked Probability Score (CRPS) \cite{gneiting2014probabilistic} and their properties, like properness, as well as benchmarking setups and competitions like the Makridakis competitions \cite{makridakis2020_m4, makridakis2022_m5}. This meta-field builds upon a broader primary field that develops methodologies for algorithms producing probabilistic forecasts, encompassing classical methods, uncertainty estimation techniques like bootstrap or conformal intervals, and modern deep learning and foundation models \cite{chen2020probabilistic, nowotarski2018probabilistic, lagllama, decoder_foundation_forecasting}.

> Despite the critical importance of evaluating probabilistic forecasts in various domains, including finance, energy, healthcare, and climate science, no comprehensive software framework or interface design has emerged to cover all these needs with a simple workflow or specification language. For instance, the reproducing code for the Makridakis competitions—while extensive in scope—relies on forecasts generated from disparate software interfaces. Similar issues are found in other benchmarking studies, where code availability is often limited or nonexistent. This lack of unified interfaces makes it challenging for practitioners in both industry and academia to contribute to or verify the growing body of evidence.

Reviewer: You make a statement here that "Similar issues are found in other benchmarking studies, where code availability is often limited or nonexistent."  This should include citations since you refer to other benchmarking studies, so you presumably can refer to them specifically.

Author: I have added references to two papers showing the lack of reproducibility.

To address these limitations, we present a simple, reproducible experimental workflow for evaluating probabilistic forecasting algorithms using \texttt{sktime} \cite{franz_kiraly_2024_11095261_sktime}. As of 2024, the \texttt{sktime} package provides the most comprehensive collection of time series-related algorithms in unified interfaces and stands out as the only major, non-commercially governed framework for time series forecasting.

> The key components of the reproducible benchmarking framework are:

Reviewer: Since you are referring to `sktime` specifically, I think you should state:
```suggestion
The key components of this reproducible benchmarking framework are:
```

  \item A first-order language that allows for the unambiguous specification of even complex forecasting algorithms.
  \item A unified software API for probabilistic performance metrics, covering metrics for distribution as well as interval or quantile forecasts.
  \item A standardized workflow for obtaining benchmark result tables for combinations of algorithms, metrics, and experimental setups.
\end{itemize}

To demonstrate the efficacy and ease of use of \texttt{sktime} in benchmarking probabilistic forecasters, we conducted a small study exploring the performance of various meta-algorithms (wrappers) that add prediction intervals to point forecasters. We investigated a range of forecasters, including Naive Forecasting and AutoTheta models \cite{autotheta}, along with probabilistic wrappers such as Conformal Intervals and Bootstrapping Methods. The study's goal was to evaluate the effectiveness of these combined approaches in improving prediction accuracy and reliability.

> We conducted experiments on several common datasets, including Australian electricity demand \cite{Godahewa2021Australian}, sunspot activity \cite{Godahewa2021Sunspot}, and US births \cite{Godahewa2021USBirth}. These datasets represent different time frequencies and characteristics.

Reviewer: The three references here to datasets are all getting rendered as "Godahewa *et al.* (2021) " instead of being rendered in some what that indicates they are separate references (e.g. Godahewa *et al.* (2021a), Godahewa *et al.* (2021b), etc.).  This is likely an issue that needs to be fixed with how Curvenote is rendering the references, but I bring it to the authors' attention.

Reviewer 2: Thanks @JuanCab - yes, this is an existing bug in MyST https://github.com/jupyter-book/mystmd/issues/272 - The links are still unique, but the text is duplicated. I will address this before the final proceedings are published! @benHeid - You can leave this as-is!

> Our paper is accompanied by easily reusable code, and we invite the open research and open-source communities to contribute to extending our experiments or using our code to set up their own. As is often the case in modern data science, computational power is a limiting factor, so we hope to leverage the \texttt{SciPy} conference to plan a more comprehensive set of studies.

Reviewer: I strongly contest this point that the paper has "easily reusable code" in it.  The code is missing imports and sample datasets to work with... simply put, the provided code can not be executed without substantial work by the reader.  The best solution of this would be to revise the code in the paper and possibly provide a downloadable code sample with a dataset.

Also, I do not believe "SciPy" should be printed in monospace font.

Author: I ensured that all larger code snippets work as standalone code. 

I removed the SciPy monospace

* [x] I will create a repository containing the exoerimental code. An alternative would be to include  the code in a dedicated section.

Author: Link to repository: https://github.com/sktime/code_for_paper_scipyconf24/tree/main

> The remainder of this paper is organized as follows: In Section \ref{methods}, we describe the forecasting methods and probabilistic wrappers used in our experiments. Section \ref{datasets} provides an overview of the datasets used for evaluation. In Section \ref{results}, we present the experimental results and discuss the performance of the combined approaches. Finally, in Section \ref{conclusion}, we conclude the paper and outline directions for future research.

Reviewer 2: These cross references (and others throughout your paper) are giving `??` in the built paper:
![image](https://github.com/scipy-conference/scipy_proceedings/assets/9453731/b897ff02-7a5c-4df1-a4a0-ad1cc1e2d8e9)

This is because `\ref` tries to create numbered references, but your sections are not numbered. You have two options:

1. Change `\ref` to `\nameref` - then you will have, for example, "In Section _Methods_, we describe..."

2. Enable section numbering in your `myst.yml`, resulting in "In Section _2_, we describe...":

  ```yaml
  # myst.yml
  project:
    ...
    numbering:
      headings: true
  ```

(Numbered sections may become the default later, but for now, taking this proactive step will fix your preview build.)

Reviewer 3: Thank you @fwkoch -- we should have turned that on in the default template, and will for next year. The final papers will have numbered sections.

@benHeid the second approach by @fwkoch above is the way to go! :)

Author: Done

\section{sktime and tsbootstrap for Reproducible Experiments}\label{sktime}

In this section, we summarize the key design principles used for reproducible benchmarking in \texttt{sktime}.

\subsection{Unified Interface}
In \texttt{sktime} and \texttt{tsbootstrap}, algorithms and mathematical objects are treated as first-class citizens. All objects of the same type, such as forecasters, follow the same interface. This consistency ensures that every forecaster is exchangeable with any other.

> Forecasters are objects with \texttt{fit} and \texttt{predict} methods, allowing both endogenous and exogenous data to be passed. For example, the following code specifies a forecaster, fits it, and makes predictions.

Reviewer: Given the target audience of the SciPy Proceedings extends beyond the machine learning community, I would avoid jargon like "endogenous" and "exogenous" (especially since these terms are only used here) and just say "data influenced by either by other model variables and/or by external factors" or something else that gets the point across without the jargon.

Author: I have rewritten it and put the ML/foreaster jargon in brackets so that these are still mentioned.

\begin{verbatim}
# specifying the forecasting algorithm
forecaster = NaiveForecaster(strategy="last", sp=12)

# fitting the forecaster -- forecast y into the future, 3 steps ahead
forecaster.fit(y, fh=[1, 2, 3])

# querying predictions
y_pred = forecaster.predict()
\end{verbatim}

A crucial design element in the above is that the first line can specify any forecaster - for instance, \texttt{ARIMA} or \texttt{ExponentialSmoothing},
and the rest of the code will work without modification.

> This design pattern is consistent across all forecasters in \texttt{sktime}, and called the "strategy pattern" in software engineering.

Reviewer: When you say "This design pattern" (the "strategy pattern") are you referring to the use of a consistent API across all forecaster classes or to the fact that it is possible to swap different forecaster classes (yes, they are related, but what is the 'design pattern', software engineering usually is not API specific, which makes me thing it is the ability to swap different forecasters that is the design pattern)... it is unclear what you refer to .

Also, this is improper latex for a quoted string, you should use 

``strategy pattern''

Author: I agree that this explanation might be confusing. We referred to the fact that the forecasters in the user's code are exchangeable. However, this is already described in the previous sentence. Thus, I removed the sentence.

Currently, \texttt{sktime} supports the construction of 82 individual forecasters, either implemented within the package or interfaced from other packages in the ecosystem. Each forecaster is parametric, with configurations ranging from 1 to 47 parameters across different forecasters. A list of forecasters can be obtained or filtered via the \texttt{all\_estimators} utility.

> Probabilistic forecasters are tagged with the \texttt{"capability:pred\_int"} tag. Currently, 45 of the 82 forecasters support probabilistic prediction modes, such as prediction intervals, quantile predictions, or full distributional predictions:

Reviewer: The concept of "tags" is not one that exists in other Python packages that I am aware of.  It might be useful to discuss tags more throughly either here or earlier when introducing `sktime`.

Author: I added an explanation of the tag system

\begin{verbatim}
# making interval predictions
y_pred_int = forecaster.predict_interval()
# making distribution predictions
y_pred_proba = forecaster.predict_proba()
\end{verbatim}

> More details can be found in the official tutorials.

Reviewer: I would provide a hyperlink to the official tutorials (I found the documentation with links to video tutorials and notebooks at [https://www.sktime.net/en/latest/users.html](https://www.sktime.net/en/latest/users.html)).

Author: Thank you. Yes this is the correct link. I added it.

> Creating algorithms with compatible APIs is straightforward.

Reviewer: I don't believe there is a need for a blank line between this sentence and the next paragraph.  This sentence can function as the lead into that paragraph.

Author: I agree, I removed the line

> \texttt{sktime} and \texttt{tsbootstrap} provide fill-in extension templates for creating algorithms in a compatible interface, which can be shared in a third-party code base and indexed by \texttt{sktime}, or contributed directly to \texttt{sktime}.

Reviewer: Can you provide a hyperlink to the "fill-in extension templates" in the paper?

Author: I added the link to https://github.com/sktime/sktime/tree/main/extension_templates

\subsection{Reproducible Specification Language}
All objects in \texttt{sktime} are uniquely specified by their construction string, which serves as a reproducible blueprint. Algorithms are intended to be stable and uniquely referenced across versions; full replicability can be achieved by freezing Python environment versions and setting random seeds.

For example, the specification string 
\texttt{NaiveForecaster(strategy="last", sp=12)}
uniquely specifies a native implementation of the seasonal last-value-carried-forecaster, with seasonality parameter 12.

Typical \texttt{sktime} code specifies the forecaster as Python code, but it can also be used as a string to store or share specifications. The \texttt{registry.craft} utility converts a Python string into an \texttt{sktime} object, ensuring easy reproducibility:

\begin{verbatim}
from sktime.registry import craft
forecaster = craft('NaiveForecaster(strategy="last", sp=12)')
\end{verbatim}

> which can be used to copy a full specification from a research paper and immediately construct the respective algorithm in a python environment, even if full code has not been shared by the author.

Reviewer: "Python" should be upper-case as it is a proper name for the programming language.

Author: Done

> which can be used to copy a full specification from a research paper and immediately construct the respective algorithm in a python environment, even if full code has not been shared by the author.

Reviewer: I didn't quite understand the importance of this `craft` feature.  It looks like it just takes a string and parses it into a function call.  What is the advantage of 
```
forecaster = NaiveForecaster(strategy="last", sp=12)
```
versus
```
forecaster = craft('NaiveForecaster(strategy="last", sp=12)')
```
They both end up creating the same object, right?  I think this needs to be clarified since I am clearly missing how this lets you "construct the respective algorithm [] even if full code has not been shared by the author."

Author: See comment here: https://github.com/scipy-conference/scipy_proceedings/pull/915#discussion_r1703283975

This approach makes it extremely easy to share specifications reproducibly, simplifying the often convoluted process of describing algorithms in research papers.

\subsection{Meta-Algorithms and Composability}
\texttt{sktime} provides not only simple algorithms as objects in unified interfaces but also meta-algorithms, such as data processing pipelines or algorithms that add interval forecast capability to existing forecasting algorithms. Importantly, these composites also follow unified interfaces.

This leads to a compositional specification language with rich combinatorial flexibility. For example, the following code adds conformal prediction interval estimates to the \texttt{NaiveForecaster}:

\begin{verbatim}
ConformalIntervals(NaiveForecaster(strategy="last", sp=12))
\end{verbatim}

> This method uses a window sliding and residual collection strategy.

Reviewer: Jargon: please define or use alternative phrasing for "window sliding" and "residual collection strategy".

Author: I tried to clarified it.

The resulting algorithm possesses \texttt{fit}, \texttt{predict}, and - added by the wrapper - \texttt{predict\_interval}, as well as the \texttt{capability:pred\_int} tag.

Data transformation pipelines can be constructed similarly, or with an operator-based specification syntax:

\begin{verbatim}
Differencer() * Deseasonalizer(sp=12) * NaiveForecaster(strategy="last")
\end{verbatim}

> This creates a forecasting algorithm that first computes first differences, then deseasonalizes with periodicity 12, then applies the last-value-carry-forward method, reseasonalizes, and inverts the first differences.

Reviewer: This sentence is full of jargon "deseasonalize" and "reseasonlize" (which I can sort of understand from context) and the "last-value-carry-forward method" which being a specific method should not only be described but probably linked to a reference via a citation.

Author: `last-value-carry-forward method` just means that the last value is used as a prediction. I clarified it and also tried to mitigate jargon.

\begin{verbatim}
y_pred_proba = forecaster.predict_proba()
crps = CRPS()

crps(y_true, y_pred_proba)
\end{verbatim}

Tags control the type of probabilistic prediction expected, such as \texttt{"scitype:y\_pred"} with the value \texttt{"pred\_proba"} for CRPS.

\subsection{Benchmarking and Evaluation}
Our benchmarking framework involves a standardized workflow for obtaining benchmark result tables for combinations of algorithms, metrics, and experimental setups. Here is an example of how to add forecasters and tasks:

\begin{verbatim}
# Example code to add forecasters and tasks
forecaster = NaiveForecaster(strategy="last", sp=12)
forecaster.fit(y_train)
y_pred = forecaster.predict(fh)
\end{verbatim}

This approach ensures that our benchmarking process is both comprehensive and reproducible. 

\begin{figure}
    \centering
    \includegraphics[width=.5\textwidth]{Figures/evaluationFramework.png}
    \caption{Benchmarking and evaluation framework}
    \label{fig:enter-label}
\end{figure}

\subsection{Prior Work}

> The design principles of \texttt{sktime} draw inspiration from several established machine learning frameworks. Notably, \texttt{scikit-learn} in \texttt{python} \cite{pedregosa2011scikit}, \texttt{mlr} in \texttt{R} \cite{bischl2016mlr}, and \texttt{Weka} \cite{hall2009weka} have pioneered the use of first-order specification languages and object-oriented design patterns for machine learning algorithms. 

Reviewer: I do not believe Python (which should be uppercase), R, or Weka should be presented in mono-spaced font as they are not packages to be imported or used within the code but rather the names of entire programming languages (or in the case of Weka, a software environment).

Author: I aggree

\texttt{sktime} extends these foundational ideas by introducing specialized interfaces for time series forecasting, including both point and probabilistic forecasts. It also incorporates unique features such as:
\begin{itemize}
    \item Probabilistic forecasters and associated metrics interfaces
    \item Reproducibility patterns and APIs
    \item Meta-algorithms for enhanced composability
    \item Inhomogeneous API composition for flexible algorithmic design
\end{itemize}

> Additionally, the \texttt{R/fable} package \cite{hyndman2020package} has contributed similar concepts specifically tailored to forecasting, which \texttt{sktime} has expanded upon or adapted to fit within its framework.

Reviewer 2: I believe you are missing bibtex entries for the following cite ids: `hyndman2020package`, `garza2022statsforecast`, `loning2019sktime`, `hyndman2018`. Can you add these?

Author: Done

> Additionally, the \texttt{R/fable} package \cite{hyndman2020package} has contributed similar concepts specifically tailored to forecasting, which \texttt{sktime} has expanded upon or adapted to fit within its framework.

Reviewer: Give it refers to it as a package, it may make sense for "R/fable" to be monospaced.  The citation "hyndman2020package" doesn't work (as noted above, you don't appear to have included this reference in your bibtex).

Author: Thank you, I added this reference now :)

By leveraging and building upon these prior works, \texttt{sktime} offers a comprehensive and adaptable toolkit for time series forecasting, fostering reproducible research and facilitating extensive benchmarking and evaluation.

> For a detailed analysis of design patterns in AI framework packages and innovations in \texttt{sktime}, see\cite{kiraly2021designing}.

Reviewer: Typo (missing space before final reference):
```suggestion
For a detailed analysis of design patterns in AI framework packages and innovations in \texttt{sktime}, see \cite{kiraly2021designing}.
```

From a software perspective, it is worth noting that \texttt{sktime} provides a unified interface across multiple packages in the time series ecosystem:

\begin{itemize}
    \item \textbf{tsbootstrap} \cite{gilda_2024_10866090_tsbootstrap, gilda2024tsbootstrap}: A library for time series bootstrapping methods, integrated with \texttt{sktime}.
    \item \textbf{statsforecast} \cite{garza2022statsforecast}: A library for statistical and econometric forecasting methods, featuring the Auto-Theta algorithm.
    \item \textbf{sktime} \cite{loning2019sktime, franz_kiraly_2024_11095261_sktime}: Contains multiple native implementations, including naive methods, all probability wrappers, and pipeline composition.
    \item \textbf{statsmodels} \cite{seabold2010statsmodels}: A foundational library for statistical methods, used for the deseasonalizer and various statistical primitives.

> \end{itemize}

Reviewer: I think it makes more sense to, in the statement the line before the `itemize` list, to state the features `sktime` provides and then mention it also provides a unified interface to 3 other second and third party packages in the bulleted list.  It just seems odd otherwise to list `sktime` as being one of the packages it offers a unified interface to...  It also makes the followup statement after the list clearer.

Also, the references to `garza2022statsforecast` and `loning2019sktime` are missing from the provided bibtex.

Author: I completely agree and moved the sktime item as described. 

I also added the missing references

> This hybrid use of \texttt{sktime} as a framework covering first-party (itself), second-party (\texttt{tsbootstrap}), and third-party (\texttt{statsmodels}, \texttt{statsforecast}) locations is significant. Credit goes to the maintainers and implementers of these packages.

Reviewer: "Credit" for what?  "Credit for this effective integration"?  "Credit for working together"?

Author: Clarified. Credit for implementing their package.

\subsection{Forecasting Pipeline}

> Each forecaster is wrapped in a \texttt{Differencer} and a \texttt{Deseasonalizer} as preprocessing steps to improve stationarity. These preprocessors are necessary because some forecasters require the time series to be stationary and non-seasonal.

Reviewer: Jargon: Explain "stationarity" or use alternative prose.

Author: I added an explanation

\begin{itemize}
    \item \texttt{Differencer}: Computes first differences, which are inverted after forecasting by cumulative summing.
    \item \texttt{Deseasonalizer(sp=data\_sp)}: Removes the seasonal component, which is added back after forecasting. It estimates the trend by applying a convolution filter to the data, removing the trend, and then averaging the de-trended series for each period to return the seasonal component.
\end{itemize}

> All forecasters are point forecasters, combined with one of the probabilistic wrappers to generate prediction intervals and probabilistic forecasts.

Reviewer: Jargon: Please define "point forecasters".

Author: I added an explanation of point forecasters

    \includegraphics[width=\textwidth]{Figures/Pipeline.png}
    \caption{The forecasting pipeline used for all the forecasters}
    \label{fig:pipeline}
\end{figure}

The partial pipeline specification, illustrated in Figure \ref{fig:pipeline}, is:

\begin{verbatim}
Differencer() * Deseasonalizer(sp=data_sp) * wrapper(forecaster)
\end{verbatim}

where variable parts are \texttt{wrapper}, \texttt{forecaster}, and \texttt{data\_sp}. These components are varied as described below.

\subsection{Component Forecasting Models}

We use several component forecasting models in this study, each with unique characteristics:

\begin{itemize}
    \item \texttt{NaiveForecaster(strategy, sp)}: Uses simple heuristics for forecasting. The \texttt{strategy} parameter allows selecting the type of algorithm:
    \begin{itemize}
        \item \texttt{mean}: Uses the mean of the last \(N\) values, with seasonal periodicity \texttt{sp} if passed.
        \item \texttt{last}: Uses the last observed, with seasonal periodicity \texttt{sp} if passed.
        \item \texttt{drift}: Fits a line between the first and last values of the considered window and extrapolates forward.
    \end{itemize}

>     \item \texttt{StatsForecastAutoTheta(sp)}: A variant of the Theta model \cite{Assimakopoulos2000} with automated parameter tuning, from the \texttt{statsforecast} library.

Reviewer: I *think* you may have meant:
```suggestion
    \item \texttt{StatsForecastAutoTheta(sp)}: A variant of the Theta model of \cite{Assimakopoulos2000} with automated parameter tuning, from the \texttt{statsforecast} library.
```

    
\end{itemize}

\subsection{Probabilistic Wrappers}
We use the following probabilistic wrappers to enhance the forecasting models:

\begin{itemize}

>     \item \texttt{ConformalIntervals(forecaster, strategy)}: Uses conformal prediction methods \cite{Stankeviciute2021} to produce non-parametric prediction intervals. Variants of the method are selected by the \texttt{strategy} parameter: \textbf{Empirical} and \textbf{Empirical Residual} use training quantiles, with the latter using symmetrized residuals. \textbf{Conformal} implements the method of \cite{Stankeviciute2021}, and \textbf{Conformal Bonferroni} applies the Bonferroni correction.

Reviewer: Can you provide a reference for the "Bonferroni correction"?

Author: I added a reference to bonferroni correction

>     \item \texttt{BaggingForecaster(bootstrap\_transformer, forecaster)}: Provides probabilistic forecasts by bootstrapping time series and aggregating the bootstrap forecasts \cite{hyndman2018, bergmeir2016}. The \texttt{BaggingForecaster} takes a bootstrap algorithm \texttt{bootstrap\_transformer}, a first-class object in \texttt{sktime}. Various bootstrap algorithms with their parameters are applied in the study.

Reviewer: The reference to `hyndman2018` is missing from bibtex.

Author: I added the reference

    \item \texttt{NaiveVariance(forecaster)}: Uses a sliding window to compute backtesting residuals, aggregated by forecasting horizon to a variance estimate. The mean is obtained from the wrapped forecaster, and variance from the pooled backtesting estimate.
    \item \texttt{SquaringResiduals(forecaster, residual\_forecaster)}: Uses backtesting residuals on the training set, squares them, and fits the \texttt{residual\_forecaster} to the squared residuals. Forecasts of \texttt{residual\_forecaster} are used as variance predictions, with mean predictions from \texttt{forecaster}, to obtain a normal distributed forecast. In this study, \texttt{residual\_forecaster} is always \texttt{NaiveForecaster(strategy="last")}.
\end{itemize}

\subsection{Bootstrapping Techniques}

> Bootstrapping methods generate multiple resampled datasets from the original time series data, which can be used as part of wrappers to estimate prediction intervals or predictive distributions. In this study, we use bootstrap algorithms from \texttt{tsbootstrap} \cite{gilda2024tsbootstrap, gilda_2024_10866090_tsbootstrap}, \texttt{sktime} \cite{franz_kiraly_2024_11095261_sktime}, and \texttt{scikit-learn} \cite{pedregosa2011scikit} compatible framework library dedicated to time series bootstrap algorithms. \texttt{sktime} adapts these algorithms via the \texttt{TSBootstrapAdapter}, used as \texttt{bootstrap\_transformer} in \texttt{BaggingForecaster}.

Reviewer: Just wanted to note this is a nice case of where you explain Bootstrapping (jargon) in line with describing what it does.  More of this when you introduce jargon would be nice to those readers from outside the machine learning discipline.

Author: Thank you. I try to clarify the other jargons similar to this.

\begin{itemize}
    \item \texttt{MovingBlockBootstrap}: Divides the time series data into overlapping blocks of a fixed size and resamples these blocks to create new datasets. The block size is chosen to capture the dependence structure in the data.

>     \item \texttt{BlockDistributionBootstrap}: Generates bootstrapped samples by fitting a distribution to the residuals of a model and then generating new residuals from the fitted distribution. This method assumes that the residuals follow a specific distribution, such as Gaussian or Poisson, and handles dependencies by resampling blocks of residuals.

Reviewer: Can you clarify, are the generated residuals added to the data to simulate real/noisy data?  I'm guessing that is the case but it is unclear.

Author: Yes, the residuals are added back to the fitted values of the model. I added a sentence to make it clearer.

\end{itemize}

In this study, these bootstrapping techniques are used to estimate the distribution of forecasts and generate robust prediction intervals and predictive distributions as part of the \texttt{BaggingForecaster}.

\subsection{Evaluation Metrics}
We evaluate the performance of our forecasting models using the following metrics:

\begin{itemize}
    
    \item \texttt{CRPS} - Continuous Ranked Probability Score \cite{matheson1976} measures the accuracy of probabilistic forecasts by comparing the predicted distribution to the observed values. The CRPS for a real-valued forecast distribution $d$ and an observation $y$ can be defined as:
    \[
    \text{CRPS}(d, y) = \mathbf{E} \left[ \left| X - y \right| \right] - \frac{1}{2} \mathbf{E} \left[ \left| X - X' \right| \right],
    \]
    where $X$ and $X'$ are independent random variables with distribution $d$.
    
    \item \texttt{PinballLoss} - the pinball loss, also known as quantile loss \cite{pinball_loss}, evaluates the accuracy of quantile forecasts by penalizing deviations from the true values based on specified quantiles.

>     For quantile forecasts $\hat{q}_1, dots, \hat{q}_k$ at levels $\tau_1, \dots, \tau_k$ and an observation $y$, the Pinball Loss is defined as:

Reviewer: Missing backslash in math:
```suggestion
    For quantile forecasts $\hat{q}_1, \dots, \hat{q}_k$ at levels $\tau_1, \dots, \tau_k$ and an observation $y$, the Pinball Loss is defined as:
```

    \begin{equation}

>     L(\hat{q}, y) = \frac{1}{k} \sum_{i=1}^k \max \left(\tau (y - \hat{q}_i), (1 - \tau_i) (\hat{q}_i - y)\right)

Reviewer: You are missing the subscript on the first tau:
```suggestion
    L(\hat{q}, y) = \frac{1}{k} \sum_{i=1}^k \max \left(\tau_i (y - \hat{q}_i), (1 - \tau_i) (\hat{q}_i - y)\right)
```

    \end{equation}

    The experiment uses the Pinball Loss at quantiles 0.05, and 0.95.

    \item \texttt{AUCalibration} - The Area between the calibration curve and the diagonal assesses how well the prediction intervals capture the true values. For observations $y_1, \dots, y_n$ and corresponding distributional predictions with quantile functions $Q_1, \dots, Q_n$ (where $Q_i = F^{-1}_i$ for the cdf $F_i$), the AUCalibration is defined as:
    \begin{equation}
    \mbox{AUC} (Q_:, y_:) = \frac{1}{N} \sum_{i=1}^N \left| c_{(i)} - \frac{i}{N}\right|,
    \end{equation}
    where $c_i := Q_i(y_i),$ and $c_{(i)}$ is the $i$-th order statistic of $c_1, \dots, c_N$.

>     \item \texttt{IntervalWidth} - width of prediction intervals, or sharpness measures the concentration of the prediction intervals. More concentrated intervals indicate higher confidence in the forecasts. Sharpness is desirable because it indicates precise predictions. Sharpness is calculated as the average width of the prediction intervals, at nominal symmetric coverage 0.9.

Reviewer: Jargon to clarify: "nominal symmetric coverage"

Author: I removed it. Since it is not required here.

>     \item \texttt{EmpiricalCoverage} - Empirical coverage measures how much of the observations are within the predicted interval. It is computed as the proportion of observations that fall within the prediction intervals at nominal coverage 0.9, providing a direct measure of the reliability of the intervals.

Reviewer: Jargon to clarify: "nominal coverage"

Author: I replaced the jargon here.

>     \item runtime. Besides metrics that assess the quality of the forecast, average runtime for an individual fit/interence run is also reported. Runtime measures the computational efficiency of the forecasting methods, which is crucial for practical applications.

Reviewer: I believe runtime should be monospaced and followed by a – here but please double check.

    \begin{tabularx}{\textwidth}{X|X|X|X|X|X|X}
         \toprule
        Dataset & Forecast Horizon & Step Width & Window Size & Cutout Period & Number of Folds & Seasonal Length \\ \midrule
        Australian Electricity Demand & 48 & 1440 & 1440 & Last Year & 12 & 48 \\ 
        Sunspot Activity & 28 & 395 & 365 & Last 40 Years & 12 & 1\\
        US Births & 28 & 395 & 365 & Whole Time Series & 12 & 1 \\
        \bottomrule
    \end{tabularx}
\end{table}

\begin{table}[h]
    \centering
    \footnotesize
    \caption{The code instantiation of the cross-validation splits used for the evaluation on the three datasets. The parameters are taken from Table \ref{tab:evaluation_splits}.}
    \label{tab:cv_splits}
    \begin{tabularx}{\textwidth}{X|X}
         \toprule
        Dataset & CV splitter \\ \midrule
        Australian Electricity Demand & \texttt{SlidingWindowSplitter(step_length=48*30, window_length=48*30,fh=range(48))} \\  
        Sunspot Activity & \texttt{SlidingWindowSplitter(step_length=395, window_length=365,fh=range(28))}\\
        US Births & \texttt{SlidingWindowSplitter(step_length=395, window_length=365,fh=range(28))}\\
        \bottomrule
    \end{tabularx}

> \end{table}

Reviewer: I suggest moving tables 1 and 2 to after the Experimental Setup section below.  They appeared here without the text to explain them before them...

Also, should the Sunspot Activity and US Births have had `step_length` of 395 or 365?  Just curious why they are different from `window_length` whereas in the Australian Electricity Demand the two parameters are the same.

Author: Thank you. I moved the tables. 

Regarding the step lengths, we selected 395 instead of 365 so that not only the same seasonality of the time series is covered in the test splits. When using a length of 365 (12 month) always the same month (of different years) would be the test split. For the Australian electricity demand dataset, the step length is unequal to the seasonality, so we used the same as for the window length

\section{Experiments} \label{experiments}
In this section, we describe the experimental setup, the datasets, the evaluation metrics, and the experimental procedures. We explain how the experiments are designed to compare the performance of different forecasting methods.

% \subsection{Software Setup}
% The software used in our experiments includes:

% \begin{itemize}
%     \item \textbf{sktime:} A library for machine learning with time series data.
%     \item \textbf{tsbootstrap:} A library for time series bootstrapping methods.
%     \item \textbf{Statsforecast:} A library for statistical and econometric forecasting methods.
%     \item \textbf{kotsu:} A library for benchmarking and validating machine learning models.
% \end{itemize}

\subsection{Experimental Setup}
To perform the benchmarking study, we use the framework described in Section~\ref{methods}. The benchmarking compares different probabilistic wrappers on different datasets and with different forecasters regarding CRPS, Pinball Loss, AUCalibration, and Runtime. 

> To enable easy replication of the experiments, we provide for each used forecaster, and wrapper the hyperparameters by providing the used python object instantiation in Table~\ref{tab:hyperparams}. These hyperparameter depend on the dataset, especially, the seasonal periodicity (sp) is chosen based on the dataset. This parameter as well as the other dataset related parameter required to create the cross validation folds used for the evaluation are described in Table~\ref{tab:evaluation_splits}. To create the cross validation folds, we use the \texttt{SlidingWindowSplitter} from \texttt{sktime}. The instantiation of the splitter for each dataset is shown in Table~\ref{tab:cv_splits}. Figure~\ref{fig:splits} is showing the resulting cross validation folds for the tree datasets.

Reviewer: Again, "Python" rather than "python" here.

Author: Done

> To enable easy replication of the experiments, we provide for each used forecaster, and wrapper the hyperparameters by providing the used python object instantiation in Table~\ref{tab:hyperparams}. These hyperparameter depend on the dataset, especially, the seasonal periodicity (sp) is chosen based on the dataset. This parameter as well as the other dataset related parameter required to create the cross validation folds used for the evaluation are described in Table~\ref{tab:evaluation_splits}. To create the cross validation folds, we use the \texttt{SlidingWindowSplitter} from \texttt{sktime}. The instantiation of the splitter for each dataset is shown in Table~\ref{tab:cv_splits}. Figure~\ref{fig:splits} is showing the resulting cross validation folds for the tree datasets.

Reviewer: There are two references to "hyperparameter" in this paragraph that seem to refer to multiple "hyperparameters".  Also, the sentence "These hyperparameter depend on the dataset, especially, the seasonal periodicity (sp) is chosen based on the dataset." reads strangely and I am not sure of your intent.  Can you clarify the prose?

Author: I have rewritten it. Hopefully, it is now easier to understand.

\begin{figure}
    \centering
    \begin{subfigure}{1\textwidth}
        \centering
        \includegraphics[width=\textwidth]{Figures/AusElecSplit.png}
        \caption{Australian Electricity Demand}
        \label{fig:aus_split}
    \end{subfigure}

    \begin{subfigure}{1\textwidth}
        \centering
        \includegraphics[width=\textwidth]{Figures/SunspotSplit.png}
        \caption{Sunspot Activity}
        \label{fig:sunspot_split}
    \end{subfigure}

    \begin{subfigure}{1\textwidth}
        \centering
        \includegraphics[width=\textwidth]{Figures/USBirthSplit.png}
        \caption{US Births}
        \label{fig:usbirth_split}

    \end{subfigure}
    \caption{The splits used for the evaluation on the three datasets. Blue indicates the training data, and orange indicates the test data. The splits are created using the parameters from Table~\ref{tab:evaluation_splits} and Table~\ref{tab:cv_splits}}
    \label{fig:splits}
\end{figure}

\subsection{Datasets} \label{datasets}
We evaluate our forecasting methods and probabilistic wrappers on several diverse time series datasets, each offering unique characteristics:

\begin{itemize}

>     \item \textbf{Australian Electricity Demand \cite{Godahewa2021Australian}}: Half-hourly electricity demand data for five series in Australia, useful for high-frequency data evaluation.

Reviewer: "five series" OF WHAT in Australia?  Not clear.

Author: Thank you for the remark. These are time series of five states of Australia. I added the five states.

    \item \textbf{Sunspot Activity \cite{Godahewa2021Sunspot}}: Weekly observations of sunspot numbers, ideal for testing forecasting robustness on long-term periodic patterns.

>     \item \textbf{US Births \cite{Godahewa2021USBirth}}: Daily birth records in the United States, with a clear seasonal pattern, suitable for daily data performance assessment.

Reviewer: I note again the three different references to data sets here render as identical.  This is likely a Curvenote issue.

\subsection{Results} \label{results}
In this section, we present the results of our experiments. We evaluate the performance of the forecasting methods combined with probabilistic wrappers on the datasets described in Section \ref{datasets}. 
%Our primary focus is on the accuracy and reliability of the prediction intervals generated by these methods.
To increase the conciseness, we calculated the rank of each probabilistic wrapper for each combination of forecaster, metric, and dataset. Afterwards, for each metric, probabilistic wrapper and dataset, we have calculated the average across all forecasters and time series. In the following, we present the results for each dataset separately. 

\subsubsection{Performance on Australian Electricity Demand}
The results for the Australian electricity demand dataset are summarized in Table \ref{table:aus_elec_results}. We compare the performance of different forecasting models and probabilistic wrappers using the previously described evaluation metrics. 

> The ranked based evaluation show that diverse results regarding the different metrics. E.g. while CI Empirical Residual performs best on CRPS, it is only mediocre regarding the Pinball Loss and the AUCalibration. On PinballLoss, the best method is CI Empirical and on AUCalibration, it is Squaring Residuals. Regarding the runtime, the fastest method is the fallback probabilistic prediction of the base forecaster. The slowest methods are NaiveVariance and SquaringResiduals. Furthermore, it seems that the ConformalIntervals are slightly faster than the BasggingForecasters.

Reviewer: You have an error, at least if I read Table 4 correctly:  " On PinballLoss, the best method is CI Empirical and on AUCalibration, it is **Moving Block Bootstrap** (not Squaring Residuals)"

Also, you clearly cut/paste the phrase "Regarding the runtime, the fastest method is the fallback probabilistic prediction of the base forecaster. The slowest methods are NaiveVariance and SquaringResiduals. Furthermore, it seems that the ConformalIntervals are slightly faster than the **BaggingForecasters**" (note the misspelling of last word) to describe the three models you fit.  This was probably because you expected to have to edit it and then discovered the results were the same in all three cases.  Since the results were the same, I would suggest cutting that text out here and using it ONCE to describe all three models afterward.

Reviewer 4: ```suggestion
The ranked based evaluation show that diverse results regarding the different metrics. E.g. while CI Empirical Residual performs best on CRPS, it is only mediocre regarding the Pinball Loss and the AUCalibration. On PinballLoss, the best method is CI Empirical and on AUCalibration, it is Squaring Residuals. Regarding the runtime, the fastest method is the fallback probabilistic prediction of the base forecaster. The slowest methods are NaiveVariance and SquaringResiduals. Furthermore, it seems that the ConformalIntervals are slightly faster than the BaggingForecasters.
```

There is also a typo on this line.

Author: Thank you for the hints. There was an error in the text. The table was correct. I fixed it.

Fallback & 9.45 & 9.66 & 5.61 & 1.00 \\
Naive Variance & 7.75 & 7.20 & 7.15 & 10.00 \\
Squaring Residuals & 7.45 & 6.20 & 5.05 & 11.00 \\
Block Distribution Bootstrap & 4.08 & 3.15 & 6.82 & 7.83 \\
Block Residual Bootstrap & 4.30 & 5.58 & 5.84 & 8.32 \\
Block Statistic Preserving Bootstrap & 4.25 & 5.87 & 6.15 & 7.29 \\
Moving Block Bootstrap & 6.12 & 6.51 & 4.26 & 6.00 \\
CI Conformal & 5.75 & 5.40 & 5.69 & 3.92 \\
CI Empirical & 3.79 & 2.63 & 6.97 & 3.30 \\
CI Empirical Residual & 2.37 & 5.42 & 5.35 & 3.49 \\
CI Bonferroni  & 10.50 & 8.05 & 6.70 & 3.70 \\
\bottomrule
\end{tabular}

\end{table}

\subsubsection{Performance on Sunspot Activity}
Table \ref{table:sunspot_results} shows the performance of our methods on the sunspot activity dataset. The long-term periodic patterns in this dataset provide a challenging test for our forecasting models.

The ranked based evaluation show that BaggingForecaster with the Block Distribution Bootstrap scores clearly best regarding the CRPS and Pinball Loss, and AUCalibration. Regarding the runtime, the fastest method is the fallback probabilistic prediction of the base forecaster. The slowest methods are NaiveVariance and SquaringResiduals. 

> Furthermore, it seems that the ConformalIntervals are slightly faster than the BasggingForecasters. 

Reviewer: Consider removing "Regarding the runtime, the fastest method is the fallback probabilistic prediction of the base forecaster. The slowest methods are NaiveVariance and SquaringResiduals.  Furthermore, it seems that the ConformalIntervals are slightly faster than the BasggingForecasters." since it is a repeat phrase (including the typo).  See previous comment.

Reviewer 4: ```suggestion
The ranked based evaluation show that BaggingForecaster with the Block Distribution Bootstrap scores clearly best regarding the CRPS and Pinball Loss, and AUCalibration. Regarding the runtime, the fastest method is the fallback probabilistic prediction of the base forecaster. The slowest methods are NaiveVariance and SquaringResiduals. 
Furthermore, it seems that the ConformalIntervals are slightly faster than the BaggingForecasters. 
```

If it is not removed... there's a typo here.

    \label{table:sunspot_results}
\begin{tabular}{lrrrr}
\toprule
Wrapper & CRPS & PinballLoss & AUCalibration & Runtime \\
\midrule
Fallback & 9.00 & 7.25 & 9.50 & 1.00 \\
Naive Variance & 6.50 & 6.25 & 7.75 & 10.00 \\
Squaring Residuals & 8.00 & 8.00 & 4.75 & 11.00 \\
Block Distribution Bootstrap & 1.00 & 1.00 & 3.00 & 7.25 \\
Block Residual Bootstrap & 6.00 & 6.25 & 5.00 & 6.75 \\
Block Statistic Preserving Bootstrap & 5.50 & 7.00 & 2.50 & 6.25 \\
Moving Block Bootstrap & 5.75 & 5.75 & 7.50 & 5.00 \\
CI Conformal & 4.75 & 4.50 & 6.75 & 4.50 \\
CI Empirical & 5.50 & 4.00 & 7.25 & 4.50 \\
CI Empirical Residual & 3.50 & 8.00 & 6.25 & 4.25 \\
CI Bonferroni  & 10.50 & 8.00 & 5.75 & 5.50 \\
\bottomrule
\end{tabular}

\end{table}

\subsubsection{Performance on US Births}
The results for the US births dataset are presented in Table \ref{table:us_births_results}. This dataset, with its clear seasonal pattern, allows us to assess the models' ability to handle daily data.
The ranked based evaluation show that BaggingForecaster with the Block Distribution Bootstrap scores best regarding the CRPS and Pinball Loss. 

> Regarding the AUCalibration, the best score is achieved by CI Conformal. Regarding the runtime, the fastest method is the fallback probabilistic prediction of the base forecaster. The slowest methods are NaiveVariance and SquaringResiduals. Furthermore, it seems that the ConformalIntervals are slightly faster than the BasggingForecasters. 

Reviewer: Consider removing "Regarding the runtime, the fastest method is the fallback probabilistic prediction of the base forecaster. The slowest methods are NaiveVariance and SquaringResiduals.  Furthermore, it seems that the ConformalIntervals are slightly faster than the BasggingForecasters." since it is a repeat phrase (including the typo).  See previous comment.

Reviewer 4: ```suggestion
Regarding the AUCalibration, the best score is achieved by CI Conformal. Regarding the runtime, the fastest method is the fallback probabilistic prediction of the base forecaster. The slowest methods are NaiveVariance and SquaringResiduals. Furthermore, it seems that the ConformalIntervals are slightly faster than the BaggingForecasters. 
```

\begin{figure}[h]
    \centering
    \includegraphics[width=\textwidth]{Figures/ARIMA_Prediction_Intervals.png}
    \caption{Prediction intervals generated by the ARIMA model}
    \label{fig:arima}
\end{figure}

\begin{figure}[h]
    \centering
    \includegraphics[width=\textwidth]{Figures/Naive_Mean_Prediction_Intervals.png}
    \caption{Prediction intervals generated by the Naive Mean model}
    \label{fig:naive_mean}
\end{figure}

\begin{figure}[h]
    \centering
    \includegraphics[width=\textwidth]{Figures/Naive_Last_Prediction_Intervals.png}
    \caption{Prediction intervals generated by the Naive Last model}
    \label{fig:naive_last}
\end{figure}

\begin{figure}[h]
    \centering
    \includegraphics[width=\textwidth]{Figures/Naive_Drift_Prediction_Intervals.png}
    \caption{Prediction intervals generated by the Naive Drift model}
    \label{fig:naive_drift}
\end{figure}
\fi

\section{Discussion and Conclusion} \label{conclusion}

> Our experiments demonstrate that the benchmarking framework in \texttt{sktime} provides an easy-to-use solution for reproducible benchmarks. We showed this by conducting simple benchmark studies of probabilistic wrappers for point forecasts on two different systems.

Reviewer: Didn't you test "three" different systems, not "two"?

Author: I agree. I changed it

Our paper is accompanied by easily reusable code, and we invite the open research and open-source communities to contribute to extending our experiments or using our code to set up their own. As is often the case in modern data science, computational power is a limiting factor, so we hope to leverage the SciPy conference to plan a more comprehensive set of studies.

The remainder of this paper is organized as follows: In Section \ref{methods}, we describe the forecasting methods and probabilistic wrappers used in our experiments. Section \ref{datasets} provides an overview of the datasets used for evaluation. In Section \ref{results}, we present the experimental results and discuss the performance of the combined approaches. Finally, in Section \ref{conclusion}, we conclude the paper and outline directions for future research.

\section{sktime and tsbootstrap for Reproducible Experiments}\label{sktime}

In this section, we summarize the key design principles used for reproducible benchmarking in \texttt{sktime}.

\subsection{Unified Interface}

> In \texttt{sktime} and \texttt{tsbootstrap}, algorithms and mathematical objects are treated as first-class citizens. All objects of the same type, such as forecasters, follow the same interface. This consistency ensures that every forecaster is exchangeable with any other.

Reviewer 4: `tsbootstrap` is not introduced in the abstract or in Section 1.

Could you please briefly introduce it and its relationship to `sktime`? It was quite distracting to have it be mentioned in this section but then:

* not used at all in the code sample in this section
* not obviously present in the link further down (*"sktime and tsbootstrap provide fill-in [extension templates](https://github.com/sktime/sktime/tree/main/extension_templates)"*)

I searched through the paper and found it is not actually defined at all in Section 2... it's linked to, with a brief description, in Section 3.

I might say something like the following in the beginning of Section 2. Or maybe even in Section 1, as `tsbootstrap` is in the title of the paper.

> *This work is aided by `tsbootstrap` ([code repository](https://github.com/astrogilda/tsbootstrap)), a Python library which provides methods to generate synthetic time-series data.*

Author: Thank you for your proposal. I added an explanation why tsbootstrap is used in the introduction.

Forecasters are objects with \texttt{fit} and \texttt{predict} methods, allowing both the target time series (endogenous data) and the time series that influence the target time series (exogenous data) to be passed. For example, the following code specifies a forecaster, fits it, and makes predictions.

\begin{verbatim}
# importing the necessary modules
from sktime.datasets import load_airline
from sktime.forecasting.naive import NaiveForecaster

# load exemplary data
y = load_airline()

# specifying the forecasting algorithm
forecaster = NaiveForecaster(strategy="last", sp=12)

# fitting the forecaster -- forecast y into the future, 3 steps ahead
forecaster.fit(y, fh=[1, 2, 3])

# querying predictions
y_pred = forecaster.predict()
\end{verbatim}

A crucial design element in the above is that the first line can specify any forecaster - for instance, \texttt{ARIMA} or \texttt{ExponentialSmoothing},
and the rest of the code will work without modification.

This design pattern is consistent across all forecasters in \texttt{sktime}, and called the "strategy pattern" in software engineering.

> Currently, \texttt{sktime} supports the construction of 82 individual forecasters, either implemented within the package or interfaced from other packages in the ecosystem. Each forecaster is parametric, with configurations ranging from 1 to 47 parameters across different forecasters. A list of forecasters can be obtained or filtered via the \texttt{all\_estimators} utility.

Reviewer 4: I do not understand what "interfaced from other packages in the ecosystem" means or where the number 82 comes from.

With `sktime==0.31.0` (the latest version as of this review), this snippet:

```python
from sktime.registry import all_estimators
print(all_estimators())
```

Returns a Python list object containing **417** classes.

At first I thought this referred to forecasters following the `sktime` API prescriptions, but distributed via other libraries. But after following the source code of `sktime.registry.all_estimators` ([code link](https://github.com/sktime/sktime/blob/7bb08dbe568237f77e4f11a837fe85d8cee41484/sktime/registry/_lookup.py#L42)), that does not seem to be the case. `all_estimators()` prints a list of Python classes which all appear to live in the `sktime` namespace.

But then again, I also see that the library imports others libraries which can produce timeseries forecasts, like `prophet` and `gluonts` ([pyproject.toml](https://github.com/sktime/sktime/blob/7bb08dbe568237f77e4f11a837fe85d8cee41484/pyproject.toml#L79)). Is the following accurate? I'd personally find it easier to understand.

> *... construction of 82 individual forecasters. Some of these are implemented directly in `sktime`, but `sktime` also provides adapters providing a unified API to forecasting routines from other libraries, such as `gluonts` and `prophet`.*

----

Similarly, I am not sure what "or filtered" in the statement about `all_estimators()` means.

I think it would be helpful to include a code snippet using `sktime.registry.all_estimators` which:

* returns the same number of forecasters as mentioned in the paragraph (for the latest version of `sktime`)
* demonstrates at least 1 example of what is meant by "filtered"

Reviewer 4: Updated comment (reading further).... the meaning of "interfaced from other packages in the ecosystem" is quite clearly explained at the beginning of Section 3 ("Algorithmic Setup").

*(the paragraph beginning "From a software perspective it is worth noting that sktime [Löning et al. (2019)](https://doi.org/10.48550/arXiv.1909.07872)[Király et al. (2024)](https://doi.org/10.5281/zenodo.11095261) contains multiple native implementations...")*

I think it'd be preferable for the language here about how many total forecasters are supported, and that language explaining the different between first-party, second-party, and third-party, to be in the same place. Section 1 would be best, I think.... `tsbootstrap` is in the title of the paper but what it actually is is not defined until the beginning of Section 3.

So, my recommendation..... move this to Section 1:

<img width="886" alt="image" src="https://github.com/user-attachments/assets/d0630ed3-c194-460b-8077-f69b55b14774">

Author: Thank you for the suggestion. It makes definitely more sense to move the explanation of first, second, third party to the introduction. 

I also added an example code for the `all_estimators`method with and without filtering.

\subsection{Reproducible Specification Language}
All objects in \texttt{sktime} are uniquely specified by their construction string, which serves as a reproducible blueprint. Algorithms are intended to be stable and uniquely referenced across versions; full replicability can be achieved by freezing Python environment versions and setting random seeds.

For example, the specification string 
\texttt{NaiveForecaster(strategy="last", sp=12)}
uniquely specifies a native implementation of the seasonal last-value-carried-forecaster, with seasonality parameter 12.

Typical \texttt{sktime} code specifies the forecaster as Python code, but it can also be used as a string to store or share specifications. The \texttt{registry.craft} utility converts a Python string into an \texttt{sktime} object, ensuring easy reproducibility:

\begin{verbatim}
from sktime.registry import craft

> forecaster = craft('NaiveForecaster(strategy="last", sp=12)')

Reviewer 4: I was very confused by this, because that string passed to `craft()` looks like it would be valid Python code. What is the purpose of `craft()` beyond avoiding the need to specify `import` statements?

This might be more impactful with a more complex example, and ideally one where the string has some obvious deviation from syntactically-valid Python code.

Or, alternatively, could you maybe show examples with and without `craft()` which generate the same forecaster in memory?

Reviewer 5: > What is the purpose of craft() beyond avoiding the need to specify import statements?

There is a high-level and low-level answer to this.

The high-level answer is, it enables reproducibility workflows by allowing the strings to be shared as artefacts, together with an environment freeze, see this tutorial:
https://github.com/sktime/sktime-workshop-pydata-prague-2023/blob/main/notebooks/05_evaluation_and_benchmarking.ipynb

The lower-level answer relates to the concept of serialization and deserialization, to make the "blueprint" unique.

* the `craft` method is a deserialization inverse to string coercion. With a pip freeze as context, this becomes a unique inverse, as the `sktime` namespace is unique. Hence `str` and `craft` form a serialization/deserialization pair, and it the usual pattern.
* as a "convenience" feature, `craft` also allow larger code blocks with variable references to define an estimator. This can also be shared as a blueprint for a complex composite.

Examples for both can be also found in the workshop notebook linked.

Further, it is a conscious design decision to remain close to syntactically valid python code, as this enables auditability by the user, which is a desideratum in reproducible research.

Reviewer: > > What is the purpose of craft() beyond avoiding the need to specify import statements?
> 
> 
> 
> There is a high-level and low-level answer to this.

Thank you for the thoughtful answer.  I was confused about this myself!

Reviewer 4: Thanks. The details you've shared here make the purpose of this function, and the reason for this functionality in `sktime`, much clearer than the example in the paper. Please do consider expanding on this in the paper, specifically in `2.2. - Reproducible Specification Language`.

The serialization format looking like like Python code led me to the (incorrect) conclusion that it literally **was** Python code. I suspect some other readers who are unfamiliar with `sktime` might also come to that come conclusion from reading this.

I think a more complex example, and a brief description of how the format intentionally looks like Python but is not actually `eval()`'d as Python, would be helpful to avoid such misunderstandings.

is fit on the time series $y_{1:t_i}$, where $t_i \neq t_j$. Afterwards, the residuals are computed on the remaining time series $y_{t_{i+1}:n}$, where $n$ is the length of the time series.
Out of these residuals, the prediction intervals are computed.
The resulting algorithm possesses \texttt{fit}, \texttt{predict}, and - added by the wrapper - \texttt{predict\_interval}, as well as the \texttt{capability:pred\_int} tag.

Data transformation pipelines can be constructed similarly, or with an operator-based specification syntax:

\begin{verbatim}
# importing the necessary modules
from sktime.datasets import load_airline
from sktime.forecasting.naive import NaiveForecaster
from sktime.transformations.series.detrend import Deseasonalizer
from sktime.transformations.series.difference import Differencer

# load exemplary data
y = load_airline()

pipeline = Differencer() * Deseasonalizer(sp=12) * NaiveForecaster(strategy="last")

# fitting the forecaster -- forecast y into the future, 3 steps ahead
pipeline.fit(y, fh=[1, 2, 3])

# making interval predictions
y_pred_int = pipeline.predict()
\end{verbatim}

> This creates a forecasting algorithm that first computes differences, then remove the seasonality (deseasonalize) by assuming a periodicity of 12, 

Reviewer 4: ```suggestion
This creates a forecasting algorithm that first computes differences, then removes the seasonality (deseasonalize) by assuming a periodicity of 12, 
```

Reviewer: I agree with the suggested correction here although to avoid the run on sentence, I suggest a period to end this.
```suggestion
This creates a forecasting algorithm that first computes differences, then removes the seasonality (deseasonalize) by assuming a periodicity of 12.
```

> then the \texttt{NaiveForecasster(strategy="last")} uses the last value as prediction for the next values (last-value-carry-forward method), finally it adds the seasonality back

Reviewer 4: ```suggestion
then the \texttt{NaiveForecaster(strategy="last")} uses the last value as prediction for the next values (last-value-carry-forward method), finally it adds the seasonality back
```

# making interval predictions
y_pred_int = forecaster.predict_interval()
# making distribution predictions
y_pred_proba = forecaster.predict_proba()

# Initialise the CRPS metric
crps = CRPS()

# Calculate the CRPS
crps(y_test, y_pred_proba)
\end{verbatim}

Tags control the type of probabilistic prediction expected, such as \texttt{"scitype:y\_pred"} with the value \texttt{"pred\_proba"} for CRPS.

\subsection{Benchmarking and Evaluation}
Our benchmarking framework involves a standardized workflow for obtaining benchmark result tables for combinations of algorithms, metrics, and experimental setups. Here is an example of how to add forecasters and tasks:

\begin{verbatim}
# Example code to add forecasters and tasks

# Import the necessary modules
from sktime.benchmarking.forecasting import ForecastingBenchmark
from sktime.datasets import load_airline
from sktime.forecasting.naive import NaiveForecaster
from sktime.performance_metrics.forecasting.probabilistic import CRPS
from sktime.split import SlidingWindowSplitter

# Initialise the benchmark object
benchmark = ForecastingBenchmark()

# Add a forecaster to the benchmark
benchmark.add_estimator(

>     estimator=NaiveForecasster(strategy="last", sp=12),

Reviewer 4: ```suggestion
    estimator=NaiveForecaster(strategy="last", sp=12),
```

Without this fix, in v0.31.0, with the following error:

```text
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
NameError: name 'NaiveForecasster' is not defined. Did you mean: 'NaiveForecaster'?
```

\section{sktime and tsbootstrap for Reproducible Experiments}\label{sktime}

In this section, we summarize the key design principles used for reproducible benchmarking in \texttt{sktime}. Thereby, from a software perspective, it is  worth noting that \textbf{sktime} \cite{loning2019sktime, franz_kiraly_2024_11095261_sktime} contains multiple native implementations, including naive methods, all probability wrappers, and pipeline composition, but also provides a unified interface across multiple packages in the time series ecosystem, e.g.:

\begin{itemize}
    \item \textbf{tsbootstrap} \cite{gilda_2024_10866090_tsbootstrap, gilda2024tsbootstrap}: A library for time series bootstrapping methods, integrated with \texttt{sktime}.
    \item \textbf{statsforecast} \cite{garza2022statsforecast}: A library for statistical and econometric forecasting methods, featuring the Auto-Theta algorithm.
    \item \textbf{statsmodels} \cite{seabold2010statsmodels}: A foundational library for statistical methods, used for the deseasonalizer and various statistical primitives.
\end{itemize}

> This hybrid use of \texttt{sktime} as a framework covering first-party (itself), second-party (\texttt{tsbootstrap}), and third-party (\texttt{statsmodels}, \texttt{statsforecast}) locations is significant.

Reviewer: I honestly don't know if the use of `locations` is grammatically correct in this sentence, I would have used `packages`.  Not demanding a correction, just a double-check to confirm this.

Author: Thanks, packages is better

Credit goes to the maintainers and implementers of these packages for implementing the contained algorithms that we can interface.

\subsection{Unified Interface}
In \texttt{sktime} and \texttt{tsbootstrap}, algorithms and mathematical objects are treated as first-class citizens. All objects of the same type, such as forecasters, follow the same interface. This consistency ensures that every forecaster is exchangeable with any other. The same holds for all bootstrapping methods in \texttt{tsbootstrap}. Thus, they are easily exchangeable enabling simple and fast experimentation.

> Forecasters are objects with \texttt{fit} and \texttt{predict} methods, allowing both the target time series (endogenous data) and the time series that influence the target time series (exogenous data) to be passed. For example, the following code specifies a forecaster, fits it, and makes predictions.

Reviewer: Much nicer!  Thanks!

\begin{verbatim}
# importing the necessary modules
from sktime.datasets import load_airline
from sktime.forecasting.naive import NaiveForecaster

# load exemplary data
y = load_airline()

# specifying the forecasting algorithm
forecaster = NaiveForecaster(strategy="last", sp=12)

# fitting the forecaster -- forecast y into the future, 3 steps ahead
forecaster.fit(y, fh=[1, 2, 3])

# querying predictions

> y_pred = forecaster.predict()

Reviewer: Thank you for adding this code, when I ran it I did get a warning (but it did complete).  The warning was:
```
~/miniconda3/envs/reviewer915/lib/python3.12/site-packages/numpy/_core/getlimits.py:227: RuntimeWarning: invalid value encountered in exp2
  epsneg_f128 = exp2(ld(-113))
 ** On entry to DLASCLS parameter number  4 had an illegal value
 ** On entry to DLASCLS parameter number  4 had an illegal value
```
Not a show stopper (and much better than what code was present before, but just noting it.

\end{verbatim}

> A crucial design element in the above is that the first line can specify any forecaster - for instance, \texttt{ARIMA} or \texttt{ExponentialSmoothing},

Reviewer: I don't think it is the "first line" any more, can I assume you mean line 9 which reads
```
forecaster = NaiveForecaster(strategy="last", sp=12)
```

Author: Thank you. You are right.

and the rest of the code will work without modification. 

Currently, \texttt{sktime} supports the construction of 82 individual forecasters.
Some of these are implemented directly in sktime, but sktime also provides adapters providing a unified API to forecasting routines from other libraries, such as gluonts and prophet.
Each forecaster is parametric, with configurations ranging from 1 to 47 parameters across different forecasters. A list of forecasters can be obtained or filtered via the \texttt{all\_estimators} utility.
\begin{verbatim}
from sktime.registry import all_estimators

# get all forecasters
all_forecasters = all_estimators("forecaster")

# get all forecasters that support prediction intervals
all_estimators("forecaster", filter_tags={"capability:pred_int": True})

\end{verbatim}

> To better filter the forecasters (as well as the other estimators such as classifiers and regressors) based on their properties and capabilities and to control their behaviour, \texttt{sktime} implements a tag system. Each estimator, such as a forecaster, has a dict \texttt{\_tags}, with a string as a key describing the name and a value of an arbitrary type describing the property.

Reviewer: Much better and clearer (especially with that code snippet), thanks!

This type system enables an easy identification of all probabilistic forecasters, since they are tagged with the \texttt{"capability:pred\_int"} tag. Currently, 45 of the 82 forecasters support probabilistic prediction modes, such as prediction intervals, quantile predictions, or full distributional predictions:

> 

Reviewer: You might want to mention you are referring to the above code snippet or move that code snippet down here and directly refer to it more inline.

Author: I referred to the previous example

\begin{verbatim}
# importing the necessary modules
from sktime.datasets import load_airline
from sktime.forecasting.naive import NaiveForecaster

# load exemplary data
y = load_airline()

# specifying the forecasting algorithm
forecaster = NaiveForecaster(strategy="last", sp=12)

# fitting the forecaster -- forecast y into the future, 3 steps ahead
forecaster.fit(y, fh=[1, 2, 3])

# making interval predictions
y_pred_int = forecaster.predict_interval()
# making distribution predictions

> y_pred_proba = forecaster.predict_proba()

Reviewer: Nice to have fully functional code, but I am realizing the code listings need captions otherwise it is hard to see what they connect to.

Author: I agree. I added replaced the environments and added captions

\subsection{Meta-Algorithms and Composability}
\texttt{sktime} provides not only simple algorithms as objects in unified interfaces but also meta-algorithms, such as data processing pipelines or algorithms that add interval forecast capability to existing forecasting algorithms. Importantly, these composites also follow unified interfaces.

This leads to a compositional specification language with rich combinatorial flexibility. For example, the following code adds conformal prediction interval estimates to the \texttt{NaiveForecaster}:

\begin{verbatim}
# importing the necessary modules
from sktime.datasets import load_airline
from sktime.forecasting.naive import NaiveForecaster
from sktime.forecasting.conformal import ConformalIntervals

# load exemplary data
y = load_airline()

# specifying the forecasting algorithm
forecaster = ConformalIntervals(NaiveForecaster(strategy="last", sp=12))

# fitting the forecaster -- forecast y into the future, 3 steps ahead
forecaster.fit(y, fh=[1, 2, 3])

# making interval predictions
y_pred_int = forecaster.predict_interval()
\end{verbatim}

> The conformal prediction interval fits multiple instances of the wrapped forecaster on on parts of the time series using window sliding. In particular, each forecaster's instance $i$

Reviewer: Typo (repeated word)
```suggestion
The conformal prediction interval fits multiple instances of the wrapped forecaster on parts of the time series using window sliding. In particular, each forecaster's instance $i$
```

is fit on the time series $y_{1:t_i}$, where $t_i \neq t_j$. Afterwards, the residuals are computed on the remaining time series $y_{t_{i+1}:n}$, where $n$ is the length of the time series.
Out of these residuals, the prediction intervals are computed.
The resulting algorithm possesses \texttt{fit}, \texttt{predict}, and - added by the wrapper - \texttt{predict\_interval}, as well as the \texttt{capability:pred\_int} tag.

Data transformation pipelines can be constructed similarly, or with an operator-based specification syntax:

\begin{verbatim}
# importing the necessary modules
from sktime.datasets import load_airline
from sktime.forecasting.naive import NaiveForecaster
from sktime.transformations.series.detrend import Deseasonalizer
from sktime.transformations.series.difference import Differencer

# load exemplary data
y = load_airline()

pipeline = Differencer() * Deseasonalizer(sp=12) * NaiveForecaster(strategy="last")

# fitting the forecaster -- forecast y into the future, 3 steps ahead
pipeline.fit(y, fh=[1, 2, 3])

# making interval predictions
y_pred_int = pipeline.predict()

> \end{verbatim}

Reviewer: Running this code crashed for me because
```
ModuleNotFoundError: Deseasonalizer requires package 'statsmodels' to be present in the python environment, but 'statsmodels' was not found. 'statsmodels' is a soft dependency and not included in the base sktime installation. Please run: `pip install statsmodels` to install the statsmodels package. To install all soft dependencies, run: `pip install sktime[all_extras]`
```
Not really a problem with the paper, but if you release the code separately, be sure to state a used needs to `pip install sktime[all_extras]`.

Author: Thank you for the hint. I added this in the requirements of the code

This creates a forecasting algorithm that first computes differences, then remove the seasonality (deseasonalize) by assuming a periodicity of 12, 

> then the \texttt{NaiveForecaster(strategy="last")} uses the last value as prediction for the next values (last-value-carry-forward method), finally it adds the seasonality back

Reviewer: ```suggestion
Then the \texttt{NaiveForecaster(strategy="last")} uses the last value as prediction for the next value (last-value-carry-forward method).  Finally it adds the seasonality back
```

    \item \texttt{Deseasonalizer(sp=data\_sp)}: Removes the seasonal component, which is added back after forecasting. It estimates the trend by applying a convolution filter to the data, removing the trend, and then averaging the de-trended series for each period to return the seasonal component.
\end{itemize}

All used forecasters are point forecasters, i.e., for each time step they provide one value (point), and no information about the uncertainty of the forecast. Thus, they are combined with one of the probabilistic wrappers to generate prediction intervals or quantile forecasts.

\begin{figure}
    \centering
    \includegraphics[width=\textwidth]{Figures/Pipeline.png}
    \caption{The forecasting pipeline used for all the forecasters}
    \label{fig:pipeline}
\end{figure}

The partial pipeline specification, illustrated in Figure \ref{fig:pipeline}, is:

\begin{verbatim}
Differencer() * Deseasonalizer(sp=data_sp) * wrapper(forecaster)
\end{verbatim}

where variable parts are \texttt{wrapper}, \texttt{forecaster}, and \texttt{data\_sp}. These components are varied as described below.

> Note the code for running the whole benchmark is provided in the Section~\ref{experiment}.

Reviewer: This reference rendered as `Section %s`,it should be as follows:
```suggestion
Note the code for running the whole benchmark is provided in the Section~\ref{experiments}.
```
